File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
def solution(edges):
    n = 0
    for a, b in edges:
        n = max(n, a, b)
    
    graph = [[] for _ in range(n + 1)]
    edge_info = [[0, 0] for _ in range(n + 1)]
    for a, b in edges:
        graph[a].append(b)
        edge_info[b][0] += 1
        edge_info[a][1] += 1
        
    root = -1
    for x, (in_edge, out_edge) in enumerate(edge_info):
        if in_edge == 0 and out_edge >= 2:
            root = x
            break
    
    donut, stick, eight = 0, 0, 0
    for node in graph[root]:
        visited = set()
        x = node
        is_donut = True
        
        while x not in visited:
            visited.add(x)
            in_edge, out_edge = edge_info[x]
            
            if out_edge == 0:
                stick += 1
                is_donut = False
                break
            
            if out_edge == 2:
                eight += 1
                is_donut = False
                break
            
            x = graph[x][0]
        
        if is_donut:
            donut += 1
            
    return root, donut, stick, eight

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    public int[] solution(int[][] edges) {
        int n = 0;
        for (int[] edge : edges) {
            n = Math.max(n, edge[0]);
            n = Math.max(n, edge[1]);
        }
        
        int[][] edgeInfo = new int[n + 1][2];
        for (int[] row : edgeInfo) {
            Arrays.fill(row, 0);
        }
        
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
        }
        
        int a = 0;
        int b = 0;
        for (int[] edge : edges) {
            a = edge[0];
            b = edge[1];
            graph.get(a).add(b);
            
            edgeInfo[a][1] += 1;
            edgeInfo[b][0] += 1;
        }
        
        int root = 0;
        for (int[] row : edgeInfo) {
            if (row[0] == 0 && row[1] >= 2) {
                break;
            }
            root++;
        }
        
        int donut = 0;
        int stick = 0;
        int eight = 0;
        for (int node : graph.get(root)) {
            Set<Integer> visited = new HashSet<>();
            int x = node;
            boolean isDonut = true;
            
            while (!visited.contains(x)) {
                visited.add(x);
                
                int in = edgeInfo[x][0];
                int out = edgeInfo[x][1];
                
                if (out == 0) {
                    stick++;
                    isDonut = false;
                    break;
                }
                if (out == 2) {
                    eight++;
                    isDonut = false;
                    break;
                }
                
                x = graph.get(x).get(0);
            }
            
            if (isDonut) {
                donut++;
            }
        }
        
        return new int[] {root, donut, stick, eight};
    }
}